### DISCUSSION TITLE
Next JS Export to static HTML always redirect to home page “/” if page refresh

### DISCUSSION THREAD

**Discussion by yvescoinpass** [vercel/next.js#10522] [Help]
# Bug report

## Describe the bug
I am creating a blogging application using Next JS and React. I already understand the basics because of this tutorial https://nextjs.org/learn/basics/getting-started.

So I tried to implement the blogging application. When I try to test it in development mode `yarn dev` the routing are fine. But if I build it `yarn build` and export into static HTML `yarn export`. The routing misses up when I click on the link it goes on the right page but when I refresh the browser it always takes me back to the `/` page.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Follow this tutorial https://nextjs.org/learn/basics/deploying-a-nextjs-app. But use the "Deploying to Your Own Environment" steps

2. Test it on development mode `yarn dev` and visit the links and refresh each page. It will not redirect to `/` page. 

3. Build it `next build`

4. Export it `next export`

5. Upload it into your own server in my case I use nginx local server or use this package https://www.npmjs.com/package/serve

6. Visit the links and refresh each page. It will always redirect to `/` page.

## Expected behavior

It should not go back to the `/` page when user refresh the page eg. pressing F5 on the browser rather stays on the current page and load the current data of the page.

## System information

- OS: Debian 9 Linux
- Browser: Goog Chrome
- Version of Next.js: "next": "^9.2.1",

**Comment by wulei7528** [comment]
I found the same problem like you！I tried to modify next.config.js by add exportPathMap,but generate ***.html instead of ***/index.html like the document！

**Reply by will0225** [reply]
I am facing the same issue now
Any help?

**Reply by nishaaanth2** [reply]
`  trailingSlash: true,
`

add it in next.config.js

**Reply by lojithv** [reply]
@nishaaanth2 Thanks a lot :)

**Reply by krishnajhapate** [reply]
In my case not working

**Reply by hermescortesm** [reply]
Thanks it works for me :D

**Reply by pascalloria** [reply]
Thanks you help me a lot :)

**Comment by AmericanWebbery** [comment]
Are you trying to run this on something like Apache? If so, I believe you have to make a tweak in your .htaccess file.

Options -MultiViews
    RewriteEngine On
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteRule ^ index.html [QSA,L]

**Reply by will0225** [reply]
I tried with .htaccess
But No luck

**Reply by shashwat-kalpvaig** [reply]
@will0225  you can try my example I have tested it with cPanel, maybe that's the solution you are looking for.

**Reply by will0225** [reply]
where can I try with your sample?

**Reply by will0225** [reply]
I mean, where is your example?

**Reply by cervmi** [reply]
You have to setup eportPathMap. Then it works: https://nextjs.org/docs/api-reference/next.config.js/exportPathMap

**Comment by yvescoinpass** [comment]
In my case I use nginx local server or use this package https://www.npmjs.com/package/serve

**Comment by AmericanWebbery** [comment]
> In my case I use nginx local server or use this package https://www.npmjs.com/package/serve

Hmm, are you hosting the 'out' folder that contains the exported site?

**Reply by MannanAnxari** [reply]
yes

**Comment by alfanzain** [comment]
> > In my case I use nginx local server or use this package https://www.npmjs.com/package/serve
> 
> Hmm, are you hosting the 'out' folder that contains the exported site?

I think, yes. I've found this problem since November 2019 and there is no cure for this yet!

**Comment by yvescoinpass** [comment]
This is sad. I am currently implementing server side rendering using this package and upload it to the server specifically Aws Amplify.

**Comment by yvescoinpass** [comment]
Does this mean we really need to use `now` setup and Zeit hosting to use Next JS?

**Reply by xessam** [reply]
hmm

**Comment by timneutkens** [comment]
Sounds like a wrong AWS Amplify configuration. You can put `next export` output on any static file server really.

**Reply by OmarZeidan** [reply]
I don't completely agree unless I am doing something wrong @timneutkens 

I ran `next export`, deployed to S3 bucket .. but whenever you refresh on a page it gives a 404.
I fixed the 404 error by falling back to `index.html` but still that is not helping.

Any thoughts? 

btw, just for the records
its working perfectly find on my local `out` folder running `❯ python -m SimpleHTTPServer`  but not the crazy S3 bucket :(

**Reply by jonasgroendahl** [reply]
Did you solve this problem for s3?

**Reply by shorn** [reply]
I am currently investigating hosting a Next.js site on S3 served via Cloudfront.
This answer here might help: https://stackoverflow.com/a/63609112/924597

Please note that I am very new to Next.js - the linked answer might be the wrong way to do things.

**Comment by alfanzain** [comment]
@timneutkens I put ```next export``` output on apache server and I can't access [deleted url]
I must go to the homepage then click the card to open it.

**Reply by zhengwei16** [reply]
shit

**Reply by deej919** [reply]
Hi, 
solution for this?

**Reply by alfanzain** [reply]
Sorry, already kicked from company and the project was abandoned.
And still no solution for that.
I am avoiding NextJS since that.
Disclaimer: NextJS is amazing. But gave me trauma on the past. So, maybe next time.

**Reply by icyJoseph** [reply]
You know, Apache must be told how to see the files. Normally people fix this with trailing slashes. Otherwise you gotta tell Apache to resolve `/mypath` to `/mypath.html`.

If you use trailing slashes, then next.js creates actual folders with `/mypath/index.html`. Trailing slashes also means that your app links must use trailing slashes, even the `next/link` ones.

**Reply by daniel-tarsoly** [reply]
> You know, Apache must be told how to see the files. Normally people fix this with trailing slashes. Otherwise you gotta tell Apache to resolve `/mypath` to `/mypath.html`.
> 
> If you use trailing slashes, then next.js creates actual folders with `/mypath/index.html`. Trailing slashes also means that your app links must use trailing slashes, even the `next/link` ones.

That was indeed the problem in my case!
This one-liner fixed the issue for me: https://nextjs.org/docs/api-reference/next.config.js/exportPathMap#adding-a-trailing-slash

**Reply by MannanAnxari** [reply]
> Sorry, already kicked from company and the project was abandoned. And still no solution for that. I am avoiding NextJS since that. Disclaimer: NextJS is amazing. But gave me trauma on the past. So, maybe next time.

bro create a .htaccess file and put this code :

**Reply by MannanAnxari** [reply]
> > Sorry, already kicked from company and the project was abandoned. And still no solution for that. I am avoiding NextJS since that. Disclaimer: NextJS is amazing. But gave me trauma on the past. So, maybe next time.
> 
> bro create a .htaccess file and put this code :

RewriteEngine On
RewriteRule ^([^/]+)/$ $1.html
RewriteCond %{REQUEST_FILENAME} !-f
RewriteRule ^([^/]+)/$ $1.html
RewriteRule ^([^/]+)/([^/]+)/$ /$1/$2.html
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_URI} !(\.[a-zA-Z0-9]{1,5}|/)$
RewriteRule (.*)$ /$1/ [R=301,L]


i have a same issue and at the end i solved this and after creating .htaccess made some changes in next.config

my next.config look like that: 

/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: false,
  trailingSlash: true,
  images: {
    unoptimized: true
  },

  exportPathMap: async function (
    defaultPathMap,
    { dev, dir, outDir, distDir, buildId }
  ) {
    return {
      '/': { page: '/' },
      '/about': { page: '/about' },
    }
  },
}

module.exports = nextConfig

**Comment by timneutkens** [comment]
@alfanzain sounds like a wrong apache config where `.html` files are not served?

**Comment by yvescoinpass** [comment]
@timneutkens  Did you try to reproduce it? Even if you followed the deployment documentation this issue will come up.

**Comment by AmericanWebbery** [comment]
You are not tied to Now to host. I am running Next JS on Netlify.

**Reply by muhammadsoleh77** [reply]
because netlify deploy using npm run build, not export. But if I deploying with npm run export. this issue still work always redirect to home page.

I have same problem with this... anyone can help?

**Comment by yvescoinpass** [comment]
@AmericanWebbery  How did you deploy? Are you using the static html deployment? Can you give us some steps?

**Comment by AmericanWebbery** [comment]
Using Netlify I have it do auto deploys from a private Github repo (it's one of the default options). I just had to set the build script for it to use and anytime I do a push to the repo, Netlify will deploy as well. I'm pretty sure it is still just hosting the "out" folder that nextjs produces when you npm run build or yarn build. I'm going to try running a Nextjs site on a static host soon and see if I get similar results to what you see.

**Comment by alfanzain** [comment]
@timneutkens I don't know.

**Comment by AmericanWebbery** [comment]
@yvescoinpass I went through the Next JS static export demo and when I run `$:npx serve` on the 'out' folder I get the same behavior that you are describing only on dynamic pages, pages created by returning data from an API. @yvescoinpass are the pages on your site pulling in the data dynamically? Not that it should matter but I'm curious. The current next.js site I've built is not using any dynamic pages so I'm curious if I would see the same behavior on my host.

**Comment by AmericanWebbery** [comment]
@yvescoinpass I also found this: https://nextjs.org/learn/excel/static-html-export/exporting-other-pages which for the demo made the pages render completely before exporting and no longer caused a refresh to fail. Not sure if it's similar to the issue you are having but may be worth checking out.

**Comment by yvescoinpass** [comment]
@AmericanWebbery  Yes some of the pages are getting the dynamic data using AWS Amplify Appsync backend API. But some are just plain static html page like **About Page**. I got the same behavior on all pages.

**Comment by yvescoinpass** [comment]
> @yvescoinpass I also found this: https://nextjs.org/learn/excel/static-html-export/exporting-other-pages which for the demo made the pages render completely before exporting and no longer caused a refresh to fail. Not sure if it's similar to the issue you are having but may be worth checking out.

@AmericanWebbery Someone commented in here. I think this https://github.com/zeit/next.js/issues/10522#issuecomment-586581389 he used exportPathMap on his project also and got the same result as me.

**Comment by erikdstock** [comment]
I think I'm having this issue as well when running `npx serve -s out`. My home page contains this (nextjs 9.2.2):
```
      <h3>
        <Link href="/foo/test">
          <a>foo/test (Link)</a>
        </Link>
      </h3>
      <h3>
        <a href="/foo/test">foo/test (a)</a>
      </h3>
```
Clicking on the `a` tags cause no change on click, while the `Link` tags do a client-side update without refresh.  However, on refreshing the page i see home page content again. Both link types update the location bar and refreshing does not change it back to `/` - just loads home screen content.

It _appears_ to be working on zeit, but i'm not clear why and am seeing some strange behavior there as well.

**Comment by timneutkens** [comment]
If you add `-s` it's going to rewrite all files to `index.html` which is not what you want when using Next.js.
```
      -s, --single                        Rewrite all not-found requests to `index.html`
```

**Comment by yvescoinpass** [comment]
@timneutkens Thanks, you answer the question about the issue if we use this package https://www.npmjs.com/package/serve. 

How about if serve the static html files on nginx? 
Do you know what are the configuration we need?

**Comment by salazarm** [comment]
In my case it turned out my 404 was being rendered which would redirect them. 
It was rendering 404 because API Routes aren't supported by Next export.

**Comment by Mistwell** [comment]
any htaccess for apache? 
Options -MultiViews
RewriteEngine On
RewriteCond %{REQUEST_FILENAME} !-f
RewriteRule ^ index.html [QSA,L]

This settings gives only redirect to homepage on pages another than '/'.
Moving on site is ok, copy paste urls in browser return me to homepage(.

**Reply by deej919** [reply]
same issue I a facing. What to do? Kindly help me.

**Comment by mayuratbillme** [comment]
> If you add `-s` it's going to rewrite all files to `index.html` which is not what you want when using Next.js.
> 
> ```
>       -s, --single                        Rewrite all not-found requests to `index.html`
> ```

Where do I add the `-s` in the `index.html` file?

**Reply by zhengwei16** [reply]
clever

**Comment by lmuntaner** [comment]
I solved this by adding the following to `next.config.js`:

```
module.exports = {
  exportTrailingSlash: true,
}
```

https://nextjs.org/docs/api-reference/next.config.js/exportPathMap#adding-a-trailing-slash

I guess this depends on your web server.

**Reply by cemreyavuz** [reply]
Adding this rule into `next.config.js` solved the issue for me too. However, I would like to note that if there is anyone else has the issue of css files not loading after adding this rule, you need to add a leading `/` character to your file paths. For example, updating the css file import from `... href="css/bootstrap.min.css" ...` to `... href="/css/bootstrap.min.css" ...` solved the problem of css files not loading on refresh. The same goes for `<a>` tag hrefs and other similar stuff too.

**Reply by sujeongy** [reply]
It works for me. Thanks. 👍👍
FYI, `exportTrailingSlash` ->`trailingSlash`, the option has been renamed!😎

URL: https://nextjs.org/docs/api-reference/next.config.js/trailing-slash

**Reply by abichen-980** [reply]
I did this and all works fine. But the api base url changed and it points to the same front end base url.

I have my frontend deployed dev.example.com and api in api.example.com, after the above change it points dev.example.com instead of api.example.com.

**Comment by kibolho** [comment]
+1

**Comment by TrejoCode** [comment]
### For Nginx:
1.- Configure next.config: [exportPathMap](https://nextjs.org/docs/api-reference/next.config.js/exportPathMap)
2.- Configure next.config: [Adding a trailing slash](https://nextjs.org/docs/api-reference/next.config.js/exportPathMap#adding-a-trailing-slash)
3.- Deploy a "normal" server block: [Sever Blocks](https://www.nginx.com/resources/wiki/start/topics/examples/server_blocks/)

Thank you for all

**Comment by muslu** [comment]
For basic **NGINX** solving:

```
location / {
    index index.html;
    try_files $uri $uri/ /index.html?$args;
}
```

`systemctl restart nginx`


for Turkish:
https://makdos.blog/javascript/1117/singlepage-uygulamalarda-build-sonrasi-sayfa-cagrildiginda-anasayfayi/

**Comment by xlxAciDxlx** [comment]
Has anyone found a way to solve this using AWS CloudFront and S3? We've tried adding `exportTrailingSlash`, but to no avail. Accessing the files by manually adding `index.html` (i.e `sponsors/index.html`) to the URL works as expected, but just trying to access `sponsors/` doesn't work.

We've also tried enabling the "Static Website Hosting" option in the S3 bucket and looking through various options for CloudFront, but nothing appeared to make a difference.

**Reply by shorn** [reply]
Are you sure you've configured CloudFront to access S3 through the correct endpoint?
"Static Website Hosting" only works properly if you access it through the website endpoint name.

https://stackoverflow.com/questions/63591544/next-js-how-to-make-links-work-with-exported-sites-when-hosted-on-aws-cloudfron/63609112#63609112

My site is Next.js and hosted on CloudFront and I have it working correctly now, though it took a deal of fiddling.

Also, if you're using the trailingSlash option - be wary of routes with multiple `.` separators in them: https://github.com/vercel/next.js/issues/16617

**Reply by xlxAciDxlx** [reply]
Thanks, @shorn ! I'll pass that onto my team members and let you know how it works out for us! We had come across your Stackoverflow response before, but I overlooked the Origin config part.

**Reply by rintala** [reply]
Thanks @shorn, this reference helped me deploying my static Next.js app on a S3. Exporting with "trailingSlash: true" was the fix for me, with getStaticPaths() and getStaticProps() in my dynamic component.

**Reply by Jaynam07** [reply]
Getting error on deploying Next js Prerendered static pages on AWS S3

Issue - Page works if .html is appended to the URL but on removing .html from URL it shows an error

Example -
https.www.example.com/abc.html - this will work
https.www.example.com/abc- this is not working

Tried by adding trailing slash config, but it didn't work.

Any solution or workaround?

**Comment by shashwat-kalpvaig** [comment]
I have fixed this issue by placing two .htaccess files. One inside root directory with following code
>
```
<IfModule mod_rewrite.c>

RewriteEngine on 
RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_FILENAME}\.html -f
RewriteRule ^(.*)$ $1.html [NC,L]
  
</IfModule>
```


It will basically redirects all non .html routes to respective html pages and other is for pages


```
<IfModule mod_rewrite.c>

  RewriteEngine On
  RewriteBase /page
  RewriteRule ^[pid]\.html$ - [L]
  RewriteCond %{REQUEST_FILENAME} !-f
  RewriteCond %{REQUEST_FILENAME} !-d
  RewriteCond %{REQUEST_FILENAME} !-l
  RewriteRule . /page/[pid].html [L]
  
</IfModule>
```

This will redirect all the routes inside page to [pid].html file.
I have tested it on cPanel so it should work with any apache configs with enabled mod_rewrites

**Reply by mohaimenmahi** [reply]
Thanks @50shashwat ! That worked for me!

**Reply by shashwat-kalpvaig** [reply]
Thanks. There is an update to it. If you have any exported files with dashes in it (e.g. [product-category].html , then rename it as [productcategory].html and the same folder rule will work on that too. 

Also if you have any file name and folder name same then the above rule won't work you have to follow another rule. For example, if you have a file called [news].html and also you have a folder named news and inside the news folder you have a file called [pageid].html then this [pageid].html won't get accessed even if you override it inside the folder because the rule will basically rewrite all the routes inside news folder and won't send the params to that file.

In that case, follow the main .htaccess as follows

`
<IfModule mod_rewrite.c>
RewriteEngine on 
RewriteRule ^contactus contactus.html [NC,L]
RewriteRule ^careers careers.html [NC,L]
RewriteRule ^support support.html [NC,L]

</IfModule>
`
Basically, place all the routes to their respective HTML root files and ignore the file which has the same name as the folder.

Move that HTML file inside that folder. so [news].html will have to go inside the news folder and there you can place code Index to be [news].html instead of index.html using the following .htaccecss file

`
<IfModule mod_rewrite.c>


  RewriteEngine On
  RewriteBase /news/$1
  RewriteRule ^[newsid]\.html$ - [L]
  RewriteCond %{REQUEST_FILENAME} !-f
  RewriteCond %{REQUEST_FILENAME} !-d
  RewriteCond %{REQUEST_FILENAME} !-l
  RewriteRule . /news/[newsid].html [L]
  
  DirectoryIndex news.html
  
</IfModule>
`

**Reply by will0225** [reply]
Ok, Thank you very much! :)

**Reply by mohaimenmahi** [reply]
@50shashwat thanks a lot :1st_place_medal:

**Reply by xessam** [reply]
Worked for me on Cpanel, thanks a lot, <3 @50shashwat

**Reply by htmllover** [reply]
Thank you so much @50shashwat
This first one worked for me 
<IfModule mod_rewrite.c>

RewriteEngine on 
RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_FILENAME}\.html -f
RewriteRule ^(.*)$ $1.html [NC,L]
  
</IfModule>

**Reply by mahto4you** [reply]
> Thank you so much @50shashwat This first one worked for me
> 
> RewriteEngine on RewriteCond %{REQUEST_FILENAME} !-d RewriteCond %{REQUEST_FILENAME}.html -f RewriteRule ^(.*)$ $1.html [NC,L]

yes this one is working

**Reply by Dayropo** [reply]
Lifesaver @50shashwat

**Comment by cervmi** [comment]
You have to first setup ExportPathMap
https://nextjs.org/docs/api-reference/next.config.js/exportPathMap

And setup apache server to serve path e.g. /abc/xxx  as /abc/xxx.html

**Comment by micheledisalvatore** [comment]
The problem is in the nginx implementation, I had this solution
```
  location / {
    try_files $uri $uri/ /index.html;
  }
```
and changed with
```
  location / {
    try_files $uri $uri.html $uri/ /index.html;
  }
```

As the export command generates html pages, you need to set nginx to check for them adding `$uri.html` as the page `/foo` will be point to `/foo.html`

**Reply by phucnd81** [reply]
worked for me. Thanks so much

**Reply by ghost** [reply]
I owe you a beer :-)

**Reply by rburchDev** [reply]
Life saver!

**Reply by rahilk82** [reply]
This also worked out for me. But I have used the Caddy Web Server instead of Nginx. I will add the Caddyfile config here.

:8080 {
   
    handle * {
        root * /var/www/home
        try_files {path}.html {path}
        file_server
    }
}

By default, the try_files will be set as below but if you face this refresh issue, then implement the above change.:

:8080 {
   
    handle * {
        root * /var/www/home
        try_files {path} /index.html
        file_server
    }
}

**Reply by luochyi** [reply]
thx

**Reply by giteshraheja** [reply]
Nice explaination. Thanks. Before this v14.1 used to work but with v14.2 this is mandatory.

**Reply by AA-Arnold** [reply]
👍🏼

**Comment by aXlireza** [comment]
I got simple solution but might not be very found of the source of the problem
Since my app redirect to index.js in every refresh, with the URL staying the same as before refresh but the route is actually being changed
I implemented this redirect if the route and address bar don't match
```
import { useRouter } from 'next/router'
import { useEffect } from 'react'

export default function Home() {
  const router = useRouter();
  useEffect(() => {
    if (window.location.pathname != router.pathname) {
      router.push(`/${window.location.pathname}`)
    }
  }, [])
.
.
.
```
my htaccess:
```
<IfModule mod_rewrite.c>
Options -MultiViews
RewriteEngine On
RewriteCond %{REQUEST_FILENAME} !-f
RewriteRule ^ index.html [QSA,L]
</IfModule>

```

**Reply by icyJoseph** [reply]
Hi,

This can't be right. Could you please activate `preserve logs` on Chrome Dev Tools Console, and refresh the page? What do you see?

If I understand correctly, when your app is at `/a/b/c`, a refresh takes you back to `/`, but the `url` remains at `/a/b/c`?

**Reply by aXlireza** [reply]
yup that's exactly what's going on

**Reply by NoahGrayDL** [reply]
This solution is working well for me. 
I'm using ```next export``` and AWS S3, CloudFront, and all the pages are wrapped with a DefaultLayout. 
and this useEffect part on the DefaultLayout. 
But this line ```router.push(`/${window.location.pathname}`)``` no needs slash at the first on my case because window.location.pathname already has it.

**Reply by PoyrazTahan** [reply]
"""
If I understand correctly, when your app is at /a/b/c, a refresh takes you back to /, but the url remains at /a/b/c?
"""

I have this problem on next js export serving from an s3 bucket staticly

**Comment by icyJoseph** [comment]
The reason why an SPA pointing at `/a/b/c` would go back to `/` on reload/refresh has to with how the server is handling requests to that page. 

I have for example `next export` countless apps to `one.com` and `surge.sh` without any config what so ever... 

A simple app with client side routing needs to handle `404`'s as 200's and let the client side router handle the not found case.

From reading most of the questions above, I think it is fair to say that this ought to be a server configuration issue. 

Alternatively, it might be that some other party is messing with the history object? race conditions, or bad business logic? 

Could anyone having this issue share a site that has it? or repo? logs from the server receiving the request?

`next export` is sort of `headless`, as in, there's no server, it's just a bunch of html and javascript files that you could put anywhere, and it should just work.

**Reply by deej919** [reply]
I have deployed my application of nextjs on apache and facing this same issue? What to do? Please suggest

**Comment by jroitgrund** [comment]
I don't understand how to serve the static export when using dynamic pages. For example, `pages/users/[user].js` will generate `out/users/[user].html` (or `out/users/[user]/index.html`) if you've configured `  trailingSlash: true`).

Given this folder structure, are there any web servers which will serve the right HTML file for requests to URLs like "/users/1/"? I guess it's not an easy problem to solve but it's not clear from the docs that this use case requires custom config on the server.

**Reply by aXlireza** [reply]
The users/[user].js will work as (users/steven.html, users/john.html) based on the getStaticPaths which declare all the possible usernames to replace as the [user] placeholder
![image](https://user-images.githubusercontent.com/94176144/159162264-57cb51af-2e68-4239-bf7b-a31bf3cac08b.png)
I have working with something a bit more complex to have two dynamic parts on the same url

**Reply by jroitgrund** [reply]
I want the routes to stay dynamic in the static export. I only just realized that this isn't supported - it's not listed under [unsupported features](https://nextjs.org/docs/advanced-features/static-html-export#unsupported-features), although the [supported features](https://nextjs.org/docs/advanced-features/static-html-export#supported-features) section does specify that dynamic routes only work when defining `getStaticPaths`.

I'll raise a separate feature request.

**Reply by icyJoseph** [reply]
In the absence of a server and with critical secrets, possible usage of NodeJS API's incompatible with the browser, then such a thing under export becomes rather non-trivial. You may have to hack something on 404 or something like that...

**Reply by nikenvy** [reply]
just switch from dynamic routes like [id] to query params

**Comment by phegman** [comment]
This is my solution when using nginx to support static export with dynamic routes.

```
server {
    server_name your-domain.com;
    root /var/www/your-domain.com/out;

    add_header X-Frame-Options "SAMEORIGIN";
    add_header X-XSS-Protection "1; mode=block";
    add_header X-Content-Type-Options "nosniff";

    index index.html;

    error_page 404 /404/index.html;

    charset utf-8;

    location ~ ^(\D*)([\d]+)\/?$ {
        try_files $1[id]/index.html $uri;
    }

    location ~ ^(\D*)([\d]+)\/(edit)\/?$ {
        try_files $1[id]/edit/index.html $uri;
    }

    location / {
        try_files $uri $uri.html $uri/ =404;
    }
}
```

This will support CRUD style routes for resources. `resource/1/` and `resource/1/edit/`

**Comment by vahidalvandi** [comment]
it not work

**Comment by madbonez** [comment]
My workable solution for apache. May be it would help someone. Settings for .htaccess:

```
<IfModule mod_rewrite.c>
    Options +FollowSymLinks -MultiViews

    RewriteEngine On

    RewriteCond %{REQUEST_FILENAME} -d
    RewriteRule ^(.*)/$ $1.html

    RewriteCond %{REQUEST_FILENAME}.html -f
    RewriteRule !.*\.html$ %{REQUEST_FILENAME}.html [L]

    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME}\.html -f
    RewriteRule ^(.*)$ $1.html [NC,L]

    # Protect some contents
    RewriteRule ^.*/?\.git+ - [F,L]

    RewriteCond %{ENV:HTTPS} !on
    RewriteRule ^(.*)$ https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]
</IfModule>
```

App built with static export command **next build && next export**. Next config without exportTrailingSlash (exportTrailing) option

**Comment by dalmasonto** [comment]
Since the problem comes with dynamic routes and after you have exported your Next app ie if you have a dynamic page in your pages `pages/blogs/[blog_id].tsx` if you navigated to this page using the Link like you clicked on a link on a page and it opened it, it could work well. But if you refresh the dynamic route, it will show the homepage file, that is `index.html`. 
This is problematic since if you share a blog link, it could not show the blog itself but rather the homepage.
Navigation to the homepage is caused by `.htaccess settings` that are like this: 
```
<IfModule mod_rewrite.c>
    RewriteEngine On
    RewriteBase /
    RewriteRule ^index\.html$ - [L]
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-l
    RewriteRule . /index.html [L]
</IfModule>
```
This settings are critical in your cPanel domain folder to avoid encountering 500 Error if you reload a certain page for you SPA like React js export and next js exported.

So working around this one was a bit messy since .htaccess ended up having a lot of redirects. I took a simple path to solve this error but it is not worth it as it may kill the whole idea of next js being SEO friendly. 

The resolve I used was to redirect to the right page on page load(that is refresh) from the home page. As shown below: 

```
function HomePage(props: any) {
  const router = useRouter()

  useEffect(() => {
    console.log("path: ", router.asPath)
    if(router.asPath !== "/"){
      router.push(router.asPath, router.asPath)
    }
  }, [])
  return (
    <>
      <FeaturedListingsSection />
      <Welcome />
    </>
  );
}

HomePage.PageLayout = AppWrapper
```

The above component is the component for my index page in my Next Js project. So I pushed the right path if the path does not match the home path ` / `. So if the router path is not the home path, just push to the home path. 
My approach is for a quick solve since I have not tested the SEO implication of the same but definitely its going to affect bots crawling my page. 

I used this approach since my account activation link which was on a dynamic link could not load, so `router.push('component/path')` to the component solved it.

**Comment by joaodfmota** [comment]
This is what I have done for dynamic urls work on nginx without trailing slash:

`/post/slug`

```
        location ~ ^/((?!_next).*)/(.+?)(?:/)?$ {
            try_files /$1/[...id].html $uri $uri.html $uri/ =404;
        }
        
        location / {
          try_files $uri $uri.html $uri/ =404;
          add_header Cache-Control "no-cache";
        }
```

**Comment by ikedam** [comment]
Same issue here.
https://github.com/vercel/next.js/discussions/10522#discussioncomment-2304314 worked for me.

The title of this thread is misleading. The browser url is not "/" but  home page for "/" is rendered. I believe you don't mean redirect (which means browser url is changed to "/").
Please consider updating the title if you could.

**Comment by nishaaanth2** [comment]
If your gonna build static website, i would suggest to use VITE JS.
NEXT JS is build mainly for Dynamic Website

**Reply by jmonster** [reply]
Random anecdote, but I'm building a static app, and I ported it from Next to Vite, and my payload went from 400kB to 1.6MB. So. I'm with you in spirit... but sticking with the tool that properly prunes the dependency graph

**Comment by hadiswaleh** [comment]
Use trailing slashes to true in nextjs config (Nextjs version 13)

**Reply by UTYFjs** [reply]
I tried to use this, now on routes like myapp.com/about images are not loaded because /about/images/image.jpg is now added to their path (should be /images/image.jpg).  what can be done about it?  my next jsc 14.2 version

**Comment by nicholaschun** [comment]
To use static pages for nextjs using s3/cloudfront, you need enable website hosting on the s3 bucket. Copy the Bucket website endpoint and paste that as the cloudfront origin. NOTE: you will have to make the s3 bucket public for this to work. Optionally, you can use a header referrer to restrict access to the s3 bucket.
Follow this documentation to see how you can host a static website with bucket endpoints. I hope this helps someone.. It took me a couple of hours to figure out

https://stackoverflow.com/questions/63591544/next-js-how-to-make-links-work-with-exported-sites-when-hosted-on-aws-cloudfron/63609112#63609112

https://repost.aws/knowledge-center/cloudfront-serve-static-website

**Comment by adhsafal** [comment]
Facing the same issue when deploying in aws amplify... Any idea how to resolve?

**Reply by saharJ95** [reply]
https://medium.com/@ersel_aker/cloudfront-nextjs-redirect-d7673f0122e3

This worked for me..

**Reply by mcometa** [reply]
@saharJ95 that worked for me as well with some tweaks because I use `output: export`. CF settings are same as the blog.

```
// next.config.ts

const nextConfig = {
  reactStrictMode: false,
  output: 'export',
  distDir: 'dist',
  trailingSlash: true,
}
```
```
// redirect-cf.tsx
'use client'

import { useEffect } from 'react'
import { NextPage } from 'next'
import { usePathname, useRouter } from 'next/navigation'

const RedirectCf: NextPage = () => {
  const pathname = usePathname()
  const router = useRouter()

  useEffect(() => {
    if (pathname !== '/' && pathname.match('/redirect-cf')) {
      router.replace(pathname)
    } else {
      router.replace('/')
    }
  }, [router, pathname])

  return <></>
}

export default RedirectCf
```

**Comment by MannanAnxari** [comment]
nahi bhai nahi chahiye
